views/js/portableElementRegistry/factory/factory.js

```diff
@@ -0,0 +1,214 @@
+                        self.trigger('providersloaded');
+                    }, reject);
+                });
+            },
+            getAllVersions : function getAllVersions(){
+                var all = {};
+                _.forIn(this._registry, function (versions, id){
+                    all[id] = _.map(versions, 'version');
+                });
+                return all;
+            },
+            getRuntime : function getRuntime(typeIdentifier, version){
+                var pci = this.get(typeIdentifier, version);
+                if(pci){
+                    return _.assign(pci.runtime, {
+                        id : pci.typeIdentifier,
+                        label : pci.label,
+                        baseUrl : pci.baseUrl
+                    });
+                }else{
+                    this.trigger('error', {
+                        message : 'no portable element runtime found',
+                        typeIdentifier : typeIdentifier,
+                        version : version
+                    });
+                }
+            },
+            getCreator : function getCreator(typeIdentifier, version){
+                var pci = this.get(typeIdentifier, version);
+                if(pci && pci.creator){
+                    return _.assign(pci.creator, {
+                        id : pci.typeIdentifier,
+                        label : pci.label,
+                        baseUrl : pci.baseUrl,
+                        response : pci.response
+                    });
+                }else{
+                    this.trigger('error', {
+                        message : 'no portable element runtime found',
+                        typeIdentifier : typeIdentifier,
+                        version : version
+                    });
+                }
+            },
+            getBaseUrl : function getBaseUrl(typeIdentifier, version){
+                var runtime = this.get(typeIdentifier, version);
+                if(runtime){
+                    return runtime.baseUrl;
+                }
+                return '';
+            },
+            loadRuntimes : function loadRuntimes(reload){
+                var self = this;
+                return new Promise(function(resolve, reject) {
+                    if(_loaded && !reload){
+                        resolve();
+                        self.trigger('runtimesloaded');
+                    }else{
+                        self.loadProviders().then(function(){
+
+                            var loadStack = [];
+
+                            _.each(__providers, function (provider){
+                                if(provider){//check that the provider is loaded
+                                    loadStack.push(provider.load());
+                                }
+                            });
+
+                            //performs the loadings in parallel
+                            Promise.all(loadStack).then(function (results){
```

Comment:
You should directly return this promise instead of explicit resolve/reject of stacked promises.

By example:
```javascript
                var self = this;
                var loadPromise;
                if(_loaded && !reload){
                    loadPromise = Promise.resolve();
                } else {
                    loadPromise = self.loadProviders().then(function(){

                        var loadStack = [];

                        _.each(__providers, function (provider){
                            if(provider){//check that the provider is loaded
                                loadStack.push(provider.load());
                            }
                        });

                        //performs the loadings in parallel
                        return Promise.all(loadStack).then(function (results){

                            var requireConfigAliases = {};

                            //update registry
                            self._registry = _.reduce(results, function (acc, _pcis){
                                return _.merge(acc, _pcis);
                            }, {});

                            //pre-configuring the baseUrl of the portable element's source
                            _.forIn(self._registry, function (versions, typeIdentifier){
                                //currently use latest runtime path
                                requireConfigAliases[typeIdentifier] = self.getBaseUrl(typeIdentifier);
                            });
                            _requirejs.config({paths : requireConfigAliases});

                            _loaded = true;
                        });
                    });
                }

                loadPromise
                    .then(function() {
                        self.trigger('runtimesloaded');
                    })
                    .catch(function(err) {
                        self.trigger('error', err);
                    });

                return loadPromise;
```